File: android/src/main/java/com/stario10module/DocumentBuilderWrapper.kt
package com.stario10module

import com.facebook.react.bridge.*
import com.starmicronics.stario10.starxpandcommand.*

class DocumentBuilderWrapper internal constructor(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "DocumentBuilderWrapper"
    }

    @ReactMethod
    fun init(promise: Promise) {
        val builder = DocumentBuilder()
        val identifier = InstanceManager.set(builder)

        promise.resolve(identifier)
    }

    @ReactMethod
    fun dispose(identifier: String, promise: Promise) {
        InstanceManager.remove(identifier)
        promise.resolve(0)
    }

    @ReactMethod
    fun settingTopMargin(identifier: String, height: Double, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            builder.settingTopMargin(height)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun settingBlackMark(identifier: String, enable: Boolean, position: String, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            val parameter = StarIO10ValueConverter.toPrinterBlackMarkParameter(enable, position)

            builder.settingBlackMark(parameter)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun settingLabel(identifier: String, enable: Boolean, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            val parameter = StarIO10ValueConverter.toPrinterLabelParameter(enable)

            builder.settingLabel(parameter)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun settingHoldPrint(identifier: String, enable: Boolean, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            val parameter = StarIO10ValueConverter.toPrinterHoldPrintParameter(enable)

            builder.settingHoldPrint(parameter)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun settingPrintableArea(identifier: String, width: Double, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            builder.settingPrintableArea(width)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addPrinter(identifier: String, printerIdentifier: String, promise: Promise) {
        val documentBuilder = InstanceManager.get(identifier)
        val printerBuilder = InstanceManager.get(printerIdentifier)

        if (documentBuilder is DocumentBuilder && printerBuilder is PrinterBuilder) {
            documentBuilder.addPrinter(printerBuilder)
            promise.resolve(true)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addDrawer(identifier: String, drawerIdentifier: String, promise: Promise) {
        val documentBuilder = InstanceManager.get(identifier)
        val drawerBuilder = InstanceManager.get(drawerIdentifier)

        if (documentBuilder is DocumentBuilder && drawerBuilder is DrawerBuilder) {
            documentBuilder.addDrawer(drawerBuilder)
            promise.resolve(true)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addBuzzer(identifier: String, buzzerIdentifier: String, promise: Promise) {
        val documentBuilder = InstanceManager.get(identifier)
        val buzzerBuilder = InstanceManager.get(buzzerIdentifier)

        if (documentBuilder is DocumentBuilder && buzzerBuilder is BuzzerBuilder) {
            documentBuilder.addBuzzer(buzzerBuilder)
            promise.resolve(true)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addMelodySpeaker(identifier: String, melodySpeakerIdentifier: String, promise: Promise) {
        val documentBuilder = InstanceManager.get(identifier)
        val melodySpeakerBuilder = InstanceManager.get(melodySpeakerIdentifier)

        if (documentBuilder is DocumentBuilder && melodySpeakerBuilder is MelodySpeakerBuilder) {
            documentBuilder.addMelodySpeaker(melodySpeakerBuilder)
            promise.resolve(true)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addDisplay(identifier: String, displayIdentifier: String, promise: Promise) {
        val documentBuilder = InstanceManager.get(identifier)
        val displayBuilder = InstanceManager.get(displayIdentifier)

        if (documentBuilder is DocumentBuilder && displayBuilder is DisplayBuilder) {
            documentBuilder.addDisplay(displayBuilder)
            promise.resolve(true)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }

    @ReactMethod
    fun addRaw(identifier: String, content: ReadableArray, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is DocumentBuilder) {
            builder.addRaw(StarIO10ValueConverter.toList(content))
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }
}

File: ios/StarXpandCommand/DocumentBuilderWrapper.m
#import "DocumentBuilderWrapper.h"
#import "StarObjectManager.h"
#import "StarIO10ValueConverter.h"
#import <React/RCTLog.h>

@import StarIO10;

@interface DocumentBuilderWrapper()

@property(assign, nonatomic) StarObjectManager *objManager;

@end

@implementation DocumentBuilderWrapper

- (instancetype)init
{
    self = [super init];
    if (self) {
        _objManager = StarObjectManager.sharedManager;
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

RCT_EXPORT_MODULE()

/// init(): Promise<string>
RCT_REMAP_METHOD(init,
                 createWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [[STARIO10StarXpandCommandDocumentBuilder alloc] init];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to create object.", nil);
        return;
    }
    
    NSString *objID = [_objManager add:builder];
    
    resolve(objID);
}

/// dispose(objectIdentifier: string): Promise<void>
RCT_REMAP_METHOD(dispose,
                 disposeWithObjectIdentifier:(nonnull NSString *)objID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    [_objManager remove:objID];
    
    resolve(nil);
}

RCT_REMAP_METHOD(settingTopMargin,
                 settingTopMarginWithObjectIdentifier:(nonnull NSString *)objID
                 height:(nonnull NSNumber *)height
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder settingTopMargin:height.doubleValue];

    resolve(nil);
}

RCT_REMAP_METHOD(settingBlackMark,
                 settingBlackMarkWithObjectIdentifier:(nonnull NSString *)objID
                 enable:(BOOL)enable
                 position:(nonnull NSString *)position
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandPrinterBlackMarkParameter *param = [StarIO10ValueConverter toPrinterBlackMarkParameterWithStart:enable position:position];

    [builder settingBlackMark:param];
    
    resolve(nil);
}

RCT_REMAP_METHOD(settingLabel,
                 settingLabelWithObjectIdentifier:(nonnull NSString *)objID
                 enable:(BOOL)enable
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandPrinterLabelParameter *param = [StarIO10ValueConverter toPrinterLabelParameterWithStart:enable];
    
    [builder settingLabel:param];
    
    resolve(nil);
}

RCT_REMAP_METHOD(settingHoldPrint,
                 settingHoldPrintWithObjectIdentifier:(nonnull NSString *)objID
                 enable:(BOOL)enable
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandPrinterHoldPrintParameter *param = [StarIO10ValueConverter toPrinterHoldPrintParameterWithEnable:enable];
    
    [builder settingHoldPrint:param];
    
    resolve(nil);
}

RCT_REMAP_METHOD(settingPrintableArea,
                 settingPrintableAreaWithObjectIdentifier:(nonnull NSString *)objID
                 width:(nonnull NSNumber *)width
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder settingPrintableArea:width.doubleValue];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addPrinter,
                 addPrinterWithObjectIdentifier:(nonnull NSString *)objID
                 printerBuilderIdentifier:(nonnull NSString *)printerBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandPrinterBuilder *printerRootBuilder = [_objManager getObject:printerBuilderID];
    
    if (printerRootBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addPrinter:printerRootBuilder];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addDrawer,
                 addDrawerWithObjectIdentifier:(nonnull NSString *)objID
                 drawerBuilderIdentifier:(nonnull NSString *)drawerBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandDrawerBuilder *drawerBuilder = [_objManager getObject:drawerBuilderID];
    if (drawerBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addDrawer:drawerBuilder];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addBuzzer,
                 addBuzzerWithObjectIdentifier:(nonnull NSString *)objID
                 buzzerBuilderIdentifier:(nonnull NSString *)buzzerBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandBuzzerBuilder *buzzerBuilder = [_objManager getObject:buzzerBuilderID];
    if (buzzerBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addBuzzer:buzzerBuilder];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addMelodySpeaker,
                 addMelodySpeakerWithObjectIdentifier:(nonnull NSString *)objID
                 melodySpeakerBuilderIdentifier:(nonnull NSString *)melodySpeakerBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandMelodySpeakerBuilder *melodySpeakerBuilder = [_objManager getObject:melodySpeakerBuilderID];
    if (melodySpeakerBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addMelodySpeaker:melodySpeakerBuilder];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addDisplay,
                 addDisplayWithObjectIdentifier:(nonnull NSString *)objID
                 displayBuilderIdentifier:(nonnull NSString *)displayBuilderID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandDisplayBuilder *displayBuilder = [_objManager getObject:displayBuilderID];
    if (displayBuilder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    [builder addDisplay:displayBuilder];
    
    resolve(nil);
}

RCT_REMAP_METHOD(addRaw,
                 addRawWithObjectIdentifier:(nonnull NSString *)objID
                 content:(nonnull NSArray<NSNumber *> *)content
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandDocumentBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    NSData *data = [StarIO10ValueConverter toData:content];
    
    [builder addRaw:data];
    
    resolve(nil);
}

@end
